File: backend/docker-containers/python/execute.py
#!/usr/bin/env python3
import sys
import os
import json
import subprocess
import time
import tempfile
import signal
from pathlib import Path

class CodeExecutor:
    def __init__(self):
        self.timeout = int(os.environ.get('EXECUTION_TIMEOUT', '30'))
        self.max_memory = os.environ.get('MAX_MEMORY', '128m')
        self.temp_dir = tempfile.mkdtemp(prefix='execution_')
        
    def execute_code(self, code, test_cases=None):
        """Execute Python code safely with timeout and resource limits"""
        try:
            # Write code to temporary file
            code_file = Path(self.temp_dir) / 'main.py'
            with open(code_file, 'w') as f:
                f.write(code)
            
            # Set up execution environment
            env = os.environ.copy()
            env['PYTHONPATH'] = str(self.temp_dir)
            
            # Execute with timeout
            start_time = time.time()
            try:
                result = subprocess.run(
                    [sys.executable, str(code_file)],
                    cwd=self.temp_dir,
                    capture_output=True,
                    text=True,
                    timeout=self.timeout,
                    env=env
                )
                
                execution_time = (time.time() - start_time) * 1000  # ms
                
                return {
                    'success': result.returncode == 0,
                    'stdout': result.stdout,
                    'stderr': result.stderr,
                    'exit_code': result.returncode,
                    'execution_time': execution_time
                }
                
            except subprocess.TimeoutExpired:
                return {
                    'success': False,
                    'stdout': '',
                    'stderr': f'Execution timed out after {self.timeout} seconds',
                    'exit_code': -1,
                    'execution_time': self.timeout * 1000
                }
                
        except Exception as e:
            return {
                'success': False,
                'stdout': '',
                'stderr': f'Execution error: {str(e)}',
                'exit_code': -1,
                'execution_time': 0
            }
        finally:
            self.cleanup()
    
    def cleanup(self):
        """Clean up temporary files"""
        try:
            import shutil
            shutil.rmtree(self.temp_dir, ignore_errors=True)
        except:
            pass

def main():
    """Main execution entry point"""
    try:
        # Read input from stdin or environment
        input_data = sys.stdin.read() if not sys.stdin.isatty() else '{}'
        data = json.loads(input_data) if input_data.strip() else {}
        
        code = data.get('code', '')
        test_cases = data.get('test_cases', [])
        
        if not code:
            print(json.dumps({
                'success': False,
                'error': 'No code provided'
            }))
            return
        
        executor = CodeExecutor()
        result = executor.execute_code(code, test_cases)
        
        print(json.dumps(result))
        
    except Exception as e:
        print(json.dumps({
            'success': False,
            'error': f'Executor error: {str(e)}'
        }))

if __name__ == '__main__':
    main()

File: backend/docker-containers/javascript/execute.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const { spawn } = require('child_process');
const os = require('os');

class CodeExecutor {
    constructor() {
        this.timeout = parseInt(process.env.EXECUTION_TIMEOUT || '30000'); // ms
        this.maxMemory = process.env.MAX_MEMORY || '128m';
        this.tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'execution-'));
    }

    async executeCode(code, testCases = []) {
        try {
            // Write code to temporary file
            const codeFile = path.join(this.tempDir, 'main.js');
            fs.writeFileSync(codeFile, code);

            // Execute with timeout
            const startTime = Date.now();
            
            const result = await this.runWithTimeout(codeFile);
            const executionTime = Date.now() - startTime;

            return {
                success: result.exitCode === 0,
                stdout: result.stdout,
                stderr: result.stderr,
                exit_code: result.exitCode,
                execution_time: executionTime
            };

        } catch (error) {
            return {
                success: false,
                stdout: '',
                stderr: `Execution error: ${error.message}`,
                exit_code: -1,
                execution_time: 0
            };
        } finally {
            this.cleanup();
        }
    }

    runWithTimeout(codeFile) {
        return new Promise((resolve) => {
            const child = spawn('node', [codeFile], {
                cwd: this.tempDir,
                stdio: ['pipe', 'pipe', 'pipe']
            });

            let stdout = '';
            let stderr = '';
            let timedOut = false;

            // Set timeout
            const timer = setTimeout(() => {
                timedOut = true;
                child.kill('SIGKILL');
            }, this.timeout);

            child.stdout.on('data', (data) => {
                stdout += data.toString();
            });

            child.stderr.on('data', (data) => {
                stderr += data.toString();
            });

            child.on('close', (code) => {
                clearTimeout(timer);
                
                if (timedOut) {
                    resolve({
                        exitCode: -1,
                        stdout: '',
                        stderr: `Execution timed out after ${this.timeout}ms`
                    });
                } else {
                    resolve({
                        exitCode: code || 0,
                        stdout,
                        stderr
                    });
                }
            });

            child.on('error', (error) => {
                clearTimeout(timer);
                resolve({
                    exitCode: -1,
                    stdout: '',
                    stderr: `Process error: ${error.message}`
                });
            });
        });
    }

    cleanup() {
        try {
            const fs = require('fs');
            fs.rmSync(this.tempDir, { recursive: true, force: true });
        } catch (error) {
            // Ignore cleanup errors
        }
    }
}

async function main() {
    try {
        let inputData = '';
        
        // Read from stdin
        process.stdin.setEncoding('utf8');
        
        for await (const chunk of process.stdin) {
            inputData += chunk;
        }

        const data = inputData.trim() ? JSON.parse(inputData) : {};
        const code = data.code || '';
        const testCases = data.test_cases || [];

        if (!code) {
            console.log(JSON.stringify({
                success: false,
                error: 'No code provided'
            }));
            return;
        }

        const executor = new CodeExecutor();
        const result = await executor.executeCode(code, testCases);
        
        console.log(JSON.stringify(result));

    } catch (error) {
        console.log(JSON.stringify({
            success: false,
            error: `Executor error: ${error.message}`
        }));
    }
}

main().catch(console.error);
